### PR TITLE
[SYCL] Fix depends_on handling with pi commands

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**
Fixes two related issues:
1) if pi task is blocked by host task or host accessor it can not be enqueued and piEvent is not present in its event_impl. When we schedule new pi task with explicit (depends_on) dependency on the first one - its is absent in MDeps since we have no usual memory dependencies and present in MPreparedDepsEvents. MPreparedDepsEvents is used in enqueueImp for obtaining piEvents. Any events from MPreparedDepsEvents w/o pi events will be just skipped.
AddDep always call processDepEvent which distributes events to MPreparedDepsEvents (pi event expected) and MPreparedHostDepsEvents (no pi event) so replacement of MDeps in enqueueCommand should be valid.
2) if we have kernel w/o usual memory dependencies (MDeps & MUsers are empty) blocked kernel will be just "cleanup" and its execution will be skipped.

### CODE REVIEW

**Comment by KseniyaTikhomirova** [comment]
test for depends_on is included here https://github.com/intel/llvm/pull/5876

**Comment by KseniyaTikhomirova** [comment]
@sergey-semenov , @s-kanaev , @romanovvlad , hi, folks, could you please review since you know almost all context for this change?

File: sycl/source/detail/scheduler/commands.cpp [hunk 1]
```diff
@@ -585,14 +596,14 @@ const ContextImplPtr &Command::getWorkerContext() const {
...
+  return MWorkerQueue;
+}
 
 bool Command::producesPiEvent() const { return true; }
 
-bool Command::supportsPostEnqueueCleanup() const {
-  // Isolated commands are cleaned up separately
-  return !MUsers.empty() || !MDeps.empty();
-}
+bool Command::supportsPostEnqueueCleanup() const { return true; }
```
Comment (alexbatashev) [inline sycl/source/detail/scheduler/commands.cpp:606]: This function is not virtual, and it's not doing anything useful. Why do we need it?

Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:606]: /// Returns true iff this command can be freed by post enqueue cleanup.
  virtual bool supportsPostEnqueueCleanup() const;

it is virtual, and some accessors of Command overrides it (e.g. ExecCGCommand)

Comment (alexbatashev) [reply sycl/source/detail/scheduler/commands.cpp:606]: I see, ok

File: sycl/source/detail/event_impl.hpp [hunk 2]
```diff
@@ -209,6 +209,11 @@ class event_impl {
   }
   bool needsCleanupAfterWait() { return MNeedsCleanupAfterWait; }
 
+  /// Returns worker queue for command.
+  ///
+  /// @return a reference to MWorkerQueue.
+  QueueImplPtr &getWorkerQueue() { return MWorkerQueue; };
```
Comment (s-kanaev) [inline sycl/source/detail/event_impl.hpp:215]: Is it both setter and getter method?
Couldn't find a c-tor which sets worker queue.
```suggestion
  const QueueImplPtr &getWorkerQueue() const { return MWorkerQueue; };
```

File: sycl/source/detail/scheduler/commands.cpp [hunk 3]
```diff
@@ -169,12 +169,26 @@ static std::string commandToName(Command::CommandType Type) {
...
-    if (EventImpl->getHandleRef() != nullptr)
-      RetPiEvents.push_back(EventImpl->getHandleRef());
+    if (EventImpl->getHandleRef() == nullptr)
+      continue;
+
+    // Do not add redundant event dependencies for in-order queues.
+    // At this stage dependency is definitely pi task and need to check if
+    // current one is a host task. In this case we should not skip pi event due
+    // to different sync mechanisms for different task types on in-order queue.
+    auto WorkerQueue = getWorkerQueue();
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.cpp:183]: Please, be specific with type here. Also, do we need a copy of shared_ptr here? Will a reference to one do?

File: sycl/source/detail/scheduler/commands.cpp [hunk 4]
```diff
@@ -169,12 +169,26 @@ static std::string commandToName(Command::CommandType Type) {
...
+    if (EventImpl->getWorkerQueue() == WorkerQueue &&
+        WorkerQueue->has_property<property::queue::in_order>() &&
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.cpp:185]: NIT: I reckon value of this property is cached, isn't it?

File: sycl/source/detail/scheduler/commands.cpp [hunk 5]
```diff
@@ -169,12 +169,26 @@ static std::string commandToName(Command::CommandType Type) {
...
+        (MType != CommandType::RUN_CG /* host task has this type also */ ||
+         (static_cast<const ExecCGCommand *>(this))->getCG().getType() !=
+             CG::CGTYPE::CodeplayHostTask))
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.cpp:188]: NIT: it seems like this one deserves a macro or static function as we have several places for such checks.

File: sycl/source/detail/scheduler/commands.cpp [hunk 6]
```diff
@@ -1299,10 +1311,6 @@ const ContextImplPtr &MemCpyCommand::getWorkerContext() const {
   return getWorkerQueue()->getContextImplPtr();
 }
 
-const QueueImplPtr &MemCpyCommand::getWorkerQueue() const {
-  return MQueue->is_host() ? MSrcQueue : MQueue;
-}
-
```
Comment (s-kanaev) [inline sycl/source/detail/scheduler/commands.cpp:1305]: What's reasoning for removing of this method?

**Review by s-kanaev — COMMENTED** [review]
Partial review done.
To be continued :-)

File: sycl/source/detail/event_impl.hpp (hunk 2 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/event_impl.hpp:215]: it has the same approach as with MPreparedTaskDeps. WorkerQueue can be different depending on command type and it is assigned by command in its constructor.

File: sycl/source/detail/scheduler/commands.cpp (hunk 3 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:183]: ok, will fix, const reference will be fine.

File: sycl/source/detail/scheduler/commands.cpp (hunk 4 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:185]: seems to be not, but replacing with function will be more readable (will do)

bool queue::is_in_order() const {
  return impl->has_property<property::queue::in_order>();
}

File: sycl/source/detail/scheduler/commands.cpp (hunk 6 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:1305]: MWorkerQueue is assigned in Command constructor with default value and in specific command accessors constructor reassigned to another value (if it differs from default).
that means that we can just implement method getWorkerQueue that just returns WorkerQueue - it is done and located in Command class. getWorkerQueue overloading in accessors removed.

File: sycl/source/detail/scheduler/commands.cpp (hunk 5 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:188]: will fix

File: sycl/source/detail/scheduler/commands.cpp [hunk 7]
```diff
@@ -169,12 +169,26 @@ static std::string commandToName(Command::CommandType Type) {
...
 
-static std::vector<RT::PiEvent>
-getPiEvents(const std::vector<EventImplPtr> &EventImpls) {
+std::vector<RT::PiEvent>
+Command::getPiEvents(const std::vector<EventImplPtr> &EventImpls) const {
   std::vector<RT::PiEvent> RetPiEvents;
   for (auto &EventImpl : EventImpls) {
-    if (EventImpl->getHandleRef() != nullptr)
-      RetPiEvents.push_back(EventImpl->getHandleRef());
+    if (EventImpl->getHandleRef() == nullptr)
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/commands.cpp:176]: Won't `getHandleRef` return `nullptr` for host tasks?

Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:176]: this function is used for MPreparedDepsEvents only. host tasks and MPreparedHostDepsEvents are rendered separately

File: sycl/source/detail/event_impl.cpp [hunk 8]
```diff
@@ -322,9 +322,14 @@ event_impl::get_info<info::event::command_execution_status>() const {
...
+  if (!MHostEvent) {
+    // Command is enqueued and PiEvent is ready
+    if (MEvent)
+      return get_event_info<info::event::command_execution_status>::get(
+          this->getHandleRef(), this->getPlugin());
+    // Command is blocked and not enqueued, PiEvent is not assigned yet
+    else if (MCommand)
+      return sycl::info::event_command_status::submitted;
   }
   return MHostEvent && MState.load() != HES_Complete
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/event_impl.cpp:334]: yes, now MHostEvent check seems to be redundant here, thanks.

File: sycl/source/detail/scheduler/commands.cpp (hunk 7 above)
Comment (romanovvlad) [reply sycl/source/detail/scheduler/commands.cpp:176]: Does it mean that we cannot have an event for host task here?

Comment (romanovvlad) [reply sycl/source/detail/scheduler/commands.cpp:176]: So, how then the check below(on line 187)
`(static_cast<const ExecCGCommand *>(this))->getCG().getType() != CG::CGTYPE::CodeplayHostTask))`
can pass?

Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:176]: EventsImpls that we are rendering in getPIEvents function are dependencies for "this" command and are commands isPiExpected for which is true. The check on line 187 analyzes "this" command, not dependencies. We do this check to verify that we do not have switch between sync methods here (host task event sync for in order queue vs native sync in backend).
To conclude:
"this" here can be both host task or task which produces pi event.
EventImpls here are MPreparedDepsEvents for "this" and consist only of command that produces pi events.

File: sycl/source/detail/event_impl.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/event_impl.cpp]: fixed in commit https://github.com/intel/llvm/pull/5901/commits/2b85d95db74d32a5a0593f7bf32d4b410d77fe0a

File: sycl/source/detail/scheduler/commands.cpp (hunk 5 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp:222]: fixed in commit https://github.com/intel/llvm/pull/5901/commits/a0c087b43012eb41111cefb0bf79b9e659bce928

File: sycl/source/detail/scheduler/commands.cpp (hunk 4 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp]: fixed in commit https://github.com/intel/llvm/pull/5901/commits/a0c087b43012eb41111cefb0bf79b9e659bce928

File: sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+    case sycl::detail::CG::CodeplayHostTask: {
+      CommandGroup.reset(new detail::CGHostTask(
+          std::move(getHostTask()), getQueue(), getQueue()->getContextImplPtr(),
+          getArgs(), getArgsStorage(), getAccStorage(), getSharedPtrStorage(),
+          getRequirements(), getEvents(), getCGType(), getCodeLoc()));
+      break;
+    }
+    default:
+      throw sycl::runtime_error("Unhandled type of command group",
+                                CL_INVALID_OPERATION);
```
Comment (KseniyaTikhomirova) [inline sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp:52]: to replace PI_ERROR_INVALID_OPERATION

File: sycl/unittests/scheduler/SchedulerTestUtils.hpp [hunk 10]
```diff
@@ -209,3 +209,65 @@ sycl::detail::Requirement getMockRequirement(const MemObjT &MemObj) {
...
+    static_cast<sycl::handler *>(this)->MNDRDesc.set(std::move(Range));
+  }
+
+  void addStream(const sycl::detail::StreamImplPtr &Stream) {
+    sycl::handler::addStream(Stream);
+  }
+
+  std::unique_ptr<sycl::detail::CG> finalize() {
+    throw sycl::runtime_error("Unhandled type of command group",
+                              CL_INVALID_OPERATION);
```
Comment (KseniyaTikhomirova) [inline sycl/unittests/scheduler/SchedulerTestUtils.hpp:269]: to replace PI_ERROR_INVALID_OPERATION

File: sycl/source/detail/scheduler/commands.cpp [hunk 11]
```diff
@@ -203,17 +203,34 @@ static std::string commandToName(Command::CommandType Type) {
...
+      continue;
+
+    // Do not add redundant event dependencies for in-order queues.
+    // At this stage dependency is definitely pi task and need to check if
+    // current one is a host task. In this case we should not skip pi event due
+    // to different sync mechanisms for different task types on in-order queue.
+    const auto &WorkerQueue = getWorkerQueue();
```
Comment (sergey-semenov) [inline sycl/source/detail/scheduler/commands.cpp:217]:
```suggestion
    const QueueImplPtr &WorkerQueue = getWorkerQueue();
```

**Review by sergey-semenov — APPROVED** [review]
Very minor comments aside, LGTM

File: sycl/source/detail/scheduler/commands.cpp (hunk 11 above)
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/commands.cpp]: fixed in https://github.com/intel/llvm/pull/5901/commits/7d5def94400fe92b178f43bcc778a0bd9b00d9b6

File: sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+TEST_F(SchedulerTest, InOrderEnqueueNoMemObjDoubleKernelDepHost) {
+  // Checks blocking command tranfer for dependent kernels and enqueue of root
+  // kernel on host task completion
+  unittest::ScopedEnvVar DisabledCleanup{
+      DisablePostEnqueueCleanupName, "1",
+      detail::SYCLConfig<detail::SYCL_DISABLE_POST_ENQUEUE_CLEANUP>::reset};
+
+  default_selector Selector;
+  platform Plt{Selector};
+  if (Plt.is_host()) {
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/unittests/scheduler/EnqueueWithDependsOnDeps.cpp]: fixed in https://github.com/intel/llvm/pull/5901/commits/7d5def94400fe92b178f43bcc778a0bd9b00d9b6

File: sycl/source/detail/scheduler/graph_processor.cpp [hunk 13]
```diff
@@ -58,11 +58,12 @@ bool Scheduler::GraphProcessor::enqueueCommand(
...
-  for (DepDesc &Dep : Cmd->MDeps) {
-    if (!enqueueCommand(Dep.MDepCommand, EnqueueResult, ToCleanUp, Blocking))
-      return false;
+  for (const EventImplPtr &Event : Cmd->getPreparedDepsEvents()) {
+    if (Command *DepCmd = static_cast<Command *>(Event->getCommand()))
+      if (!enqueueCommand(DepCmd, EnqueueResult, ToCleanUp, Blocking))
+        return false;
   }
 
   // Asynchronous host operations (amongst dependencies of an arbitrary command)
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/scheduler/graph_processor.cpp]: fixed in https://github.com/intel/llvm/pull/5901/commits/7d5def94400fe92b178f43bcc778a0bd9b00d9b6

**Review by sergey-semenov — APPROVED** [review]

**Comment by KseniyaTikhomirova** [comment]
@intel/llvm-gatekeepers this commit is ready to merge, could you help please?

**Pull request merged**